src/types/from_node.rs

```diff
@@ -0,0 +1,357 @@
+            interpolation: node.interpolation.into(),
+        });
+        Ok(result)
+    }
+}
+
+impl TryFrom<TransitionState> for BuiltinSpec {
+    type Error = anyhow::Error;
+
+    fn try_from(state: TransitionState) -> Result<Self, Self::Error> {
+        match state {
+            TransitionState::FixedPositionLayout(state) => state.try_into(),
+        }
+    }
+}
+
+impl From<Interpolation> for transition::Interpolation {
+    fn from(interpolation: Interpolation) -> Self {
+        match interpolation {
+            Interpolation::Linear => Self::Linear,
+            Interpolation::Spring => Self::Spring,
+        }
+    }
+}
+
+impl TryFrom<FixedPositionLayout> for BuiltinSpec {
+    type Error = anyhow::Error;
+
+    fn try_from(node: FixedPositionLayout) -> Result<Self, Self::Error> {
+        let result = Self::FixedPositionLayout(FixedPositionLayoutSpec {
+            resolution: node.resolution.into(),
+            texture_layouts: node
+                .texture_layouts
+                .into_iter()
+                .map(TryInto::try_into)
+                .collect::<Result<Vec<_>, _>>()?,
+            background_color_rgba: node
+                .background_color_rgba
+                .map(TryInto::try_into)
+                .unwrap_or(Ok(colors::RGBAColor(0, 0, 0, 0)))?,
+        });
+        Ok(result)
+    }
+}
+
+impl TryFrom<TextureLayout> for builtin_transformations::TextureLayout {
+    type Error = anyhow::Error;
+
+    fn try_from(value: TextureLayout) -> Result<Self, Self::Error> {
+        match value {
+                TextureLayout {
+                    top: None,
+                    bottom: None,
+                    ..
+                } => return Err(anyhow!("Each entry in texture_layouts in transformation \"fixed_position_layout\" requires either bottom or top coordinate.")),
+                TextureLayout {
+                    top: Some(_),
+                    bottom: Some(_),
+                    ..
+                } => return Err(anyhow!("Fields \"top\" and \"bottom\" are mutually exclusive, you can only specify one in texture layout in \"fixed_position_layout\" transformation.")),
+                _ => (),
+            };
+        match value {
+                TextureLayout {
+                    left: None,
+                    right: None,
+                    ..
+                } => return Err(anyhow!("Each entry in texture_layouts in transformation \"fixed_position_layout\" requires either right or left coordinate.")),
+                TextureLayout {
+                    left: Some(_),
+                    right: Some(_),
+                    ..
+                } => return Err(anyhow!("Fields \"left\" and \"right\" are mutually exclusive, you can only specify one in texture layout in \"fixed_position_layout\" transformation.")),
+                _ => (),
+            };
+
+        Ok(Self {
+            top: value.top.map_or(Ok(None), |v| v.try_into().map(Some))?,
```

Comment:
Not sure I understand the question?
We have a code that can both return an error and operates on optional value, so retruned value has to be a combination of Result and Option
I prefer to have Result<Option<x>> because I can use `?` to handle it.

Reply:
Ok, now I get it. I missed `?`. 

When we have type separation, I would change `builtin_transformations::TextureLayout`, such that the type system provides the guarantee, that only a single top/bottom and left/right are defined. I think we should make invalid types unrepresentable on the renderer/common level.

Reply:
yes I plan to do that, but not in this PR